ee/server/service/software_installers.go

```diff
@@ -0,0 +1,106 @@
+package service
+
+import (
+	"context"
+	"encoding/hex"
+	"strings"
+
+	"github.com/fleetdm/fleet/v4/pkg/file"
+	"github.com/fleetdm/fleet/v4/server/contexts/ctxerr"
+	"github.com/fleetdm/fleet/v4/server/fleet"
+	"github.com/go-kit/kit/log/level"
+)
+
+func (svc *Service) UploadSoftwareInstaller(ctx context.Context, payload *fleet.UploadSoftwareInstallerPayload) error {
+	if err := svc.authz.Authorize(ctx, &fleet.SoftwareInstaller{TeamID: payload.TeamID}, fleet.ActionWrite); err != nil {
+		return err
+	}
+	if payload == nil {
+		return ctxerr.New(ctx, "payload is required")
+	}
+
+	if payload.InstallerFile == nil {
+		return ctxerr.New(ctx, "installer file is required")
+	}
+
+	title, vers, hash, err := file.ExtractInstallerMetadata(payload.Filename, payload.InstallerFile)
+	if err != nil {
+		// TODO: confirm error handling
+		if strings.Contains(err.Error(), "unsupported file type") {
+			return &fleet.BadRequestError{
+				Message:     "The file should be .pkg, .msi, .exe or .deb.",
+				InternalErr: ctxerr.Wrap(ctx, err, "extracting metadata from installer"),
+			}
+		}
+		return ctxerr.Wrap(ctx, err, "extracting metadata from installer")
+	}
+	payload.Title = title
+	payload.Version = vers
+	payload.StorageID = hex.EncodeToString(hash)
+
+	// checck if exists in the installer store
+	exists, err := svc.softwareInstallStore.Exists(ctx, payload.StorageID)
+	if err != nil {
+		return ctxerr.Wrap(ctx, err, "checking if installer exists")
+	}
+	if !exists {
+		if err := svc.softwareInstallStore.Put(ctx, payload.StorageID, payload.InstallerFile); err != nil {
+			return ctxerr.Wrap(ctx, err, "storing installer")
+		}
+	}
+
+	if payload.InstallScript == "" {
+		// TODO: apply default install script
+		payload.InstallScript = "default install script"
+	}
+
+	// TODO: basic validation of install and post-install script (e.g., supported interpreters)?
+
+	// TODO: any validation of pre-install query?
```

Comment:
:+1: good points, probably some minimal validations we can apply here (similar to scripts for install/post-instal, don't know if we can do validation similar to the frontend editor of queries for the pre-install query?).